diff --git fs/bridge.go fs/bridge.go
index 96782ca..5341bc3 100644
--- fs/bridge.go
+++ fs/bridge.go
@@ -1076,15 +1076,18 @@ func (n *Inode) childrenAsDirstream() DirStream {
 	return NewListDirStream(r)
 }
 
-func (b *rawBridge) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
-	return b.readDirMaybeLookup(cancel, input, out, true)
+func (b *rawBridge) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirPlusEntryList) fuse.Status {
+	return b.readDirMaybeLookup(cancel, input, nil, out)
 }
 
-func (b *rawBridge) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
-	return b.readDirMaybeLookup(cancel, input, out, false)
+func (b *rawBridge) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirEntryList) fuse.Status {
+	return b.readDirMaybeLookup(cancel, input, out, nil)
 }
 
-func (b *rawBridge) readDirMaybeLookup(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList, lookup bool) fuse.Status {
+func (b *rawBridge) readDirMaybeLookup(cancel <-chan struct{}, input *fuse.ReadIn, outNoLookup fuse.ReadDirEntryList, outLookup fuse.ReadDirPlusEntryList) fuse.Status {
+	if (outNoLookup == nil) == (outLookup == nil) {
+		panic("Either outNoLookup or outLookup has to be set")
+	}
 	n, f := b.inode(input.NodeId, input.Fh)
 
 	direnter, ok := f.file.(FileReaddirenter)
@@ -1136,10 +1139,6 @@ func (b *rawBridge) readDirMaybeLookup(cancel <-chan struct{}, input *fuse.ReadI
 		}
 	}
 
-	defer func() {
-		f.dirOffset = out.Offset
-	}()
-
 	first := true
 	f.lastRead = f.lastRead[:0]
 	for {
@@ -1167,30 +1166,32 @@ func (b *rawBridge) readDirMaybeLookup(cancel <-chan struct{}, input *fuse.ReadI
 		if de == nil {
 			break
 		}
-
-		first = false
 		if de.Off == 0 {
 			// This logic is dup from fuse.DirEntryList, but we need the offset here so it is part of lastRead
-			de.Off = out.Offset + 1
+			de.Off = f.dirOffset + 1
 		}
-		if !lookup {
-			if !out.AddDirEntry(*de) {
+
+		first = false
+		if outLookup == nil {
+			if !outNoLookup.AddDirEntry(*de) {
 				f.overflow = *de
 				f.hasOverflow = true
 				return fuse.OK
 			}
 
 			f.lastRead = append(f.lastRead, *de)
+			f.dirOffset = de.Off
 			continue
 		}
 
-		entryOut := out.AddDirLookupEntry(*de)
+		entryOut := outLookup.AddDirLookupEntry(*de)
 		if entryOut == nil {
 			f.overflow = *de
 			f.hasOverflow = true
 			return fuse.OK
 		}
 		f.lastRead = append(f.lastRead, *de)
+		f.dirOffset = de.Off
 
 		// Virtual entries "." and ".." should be part of the
 		// directory listing, but not part of the filesystem tree.
@@ -1222,7 +1223,7 @@ func (b *rawBridge) readDirMaybeLookup(cancel <-chan struct{}, input *fuse.ReadI
 			b.setEntryOutTimeout(entryOut)
 			if de.Mode&syscall.S_IFMT != child.stableAttr.Mode&syscall.S_IFMT {
 				// The file type has changed behind our back. Use the new value.
-				out.FixMode(child.stableAttr.Mode)
+				outLookup.FixMode(child.stableAttr.Mode)
 			}
 			entryOut.Mode = child.stableAttr.Mode | (entryOut.Mode & 07777)
 		}
diff --git fuse/api.go fuse/api.go
index d1aff58..7efcf8d 100644
--- fuse/api.go
+++ fuse/api.go
@@ -466,8 +466,8 @@ type RawFileSystem interface {
 
 	// Directory handling
 	OpenDir(cancel <-chan struct{}, input *OpenIn, out *OpenOut) (status Status)
-	ReadDir(cancel <-chan struct{}, input *ReadIn, out *DirEntryList) Status
-	ReadDirPlus(cancel <-chan struct{}, input *ReadIn, out *DirEntryList) Status
+	ReadDir(cancel <-chan struct{}, input *ReadIn, out ReadDirEntryList) Status
+	ReadDirPlus(cancel <-chan struct{}, input *ReadIn, out ReadDirPlusEntryList) Status
 	ReleaseDir(input *ReleaseIn)
 	FsyncDir(cancel <-chan struct{}, input *FsyncIn) (code Status)
 
diff --git fuse/defaultraw.go fuse/defaultraw.go
index d4fb2ed..b15df5e 100644
--- fuse/defaultraw.go
+++ fuse/defaultraw.go
@@ -144,11 +144,11 @@ func (fs *defaultRawFileSystem) Fsync(cancel <-chan struct{}, input *FsyncIn) (c
 	return ENOSYS
 }
 
-func (fs *defaultRawFileSystem) ReadDir(cancel <-chan struct{}, input *ReadIn, l *DirEntryList) Status {
+func (fs *defaultRawFileSystem) ReadDir(cancel <-chan struct{}, input *ReadIn, l ReadDirEntryList) Status {
 	return ENOSYS
 }
 
-func (fs *defaultRawFileSystem) ReadDirPlus(cancel <-chan struct{}, input *ReadIn, l *DirEntryList) Status {
+func (fs *defaultRawFileSystem) ReadDirPlus(cancel <-chan struct{}, input *ReadIn, l ReadDirPlusEntryList) Status {
 	return ENOSYS
 }
 
diff --git fuse/direntry.go fuse/direntry.go
index b7d6f9f..d5ea2eb 100644
--- fuse/direntry.go
+++ fuse/direntry.go
@@ -183,3 +183,12 @@ func (l *DirEntryList) FixMode(mode uint32) {
 func (l *DirEntryList) bytes() []byte {
 	return l.buf
 }
+
+type ReadDirEntryList interface {
+	AddDirEntry(e DirEntry) bool
+}
+
+type ReadDirPlusEntryList interface {
+	AddDirLookupEntry(e DirEntry) *EntryOut
+	FixMode(mode uint32)
+}
diff --git fuse/nodefs/dir.go fuse/nodefs/dir.go
index 75cf75a..53c7917 100644
--- fuse/nodefs/dir.go
+++ fuse/nodefs/dir.go
@@ -22,7 +22,7 @@ type connectorDir struct {
 	stream []fuse.DirEntry
 }
 
-func (d *connectorDir) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) (code fuse.Status) {
+func (d *connectorDir) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirEntryList) (code fuse.Status) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
@@ -62,7 +62,7 @@ func (d *connectorDir) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out *
 	return fuse.OK
 }
 
-func (d *connectorDir) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) (code fuse.Status) {
+func (d *connectorDir) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirPlusEntryList) (code fuse.Status) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
@@ -108,6 +108,6 @@ func (d *connectorDir) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, o
 }
 
 type rawDir interface {
-	ReadDir(out *fuse.DirEntryList, input *fuse.ReadIn, c *fuse.Context) fuse.Status
-	ReadDirPlus(out *fuse.DirEntryList, input *fuse.ReadIn, c *fuse.Context) fuse.Status
+	ReadDir(out fuse.ReadDirEntryList, input *fuse.ReadIn, c *fuse.Context) fuse.Status
+	ReadDirPlus(out fuse.ReadDirPlusEntryList, input *fuse.ReadIn, c *fuse.Context) fuse.Status
 }
diff --git fuse/nodefs/fsops.go fuse/nodefs/fsops.go
index 83e07e1..18d8f8a 100644
--- fuse/nodefs/fsops.go
+++ fuse/nodefs/fsops.go
@@ -172,13 +172,13 @@ func (c *rawBridge) OpenDir(cancel <-chan struct{}, input *fuse.OpenIn, out *fus
 	return fuse.OK
 }
 
-func (c *rawBridge) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
+func (c *rawBridge) ReadDir(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirEntryList) fuse.Status {
 	node := c.toInode(input.NodeId)
 	opened := node.mount.getOpenedFile(input.Fh)
 	return opened.dir.ReadDir(cancel, input, out)
 }
 
-func (c *rawBridge) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
+func (c *rawBridge) ReadDirPlus(cancel <-chan struct{}, input *fuse.ReadIn, out fuse.ReadDirPlusEntryList) fuse.Status {
 	node := c.toInode(input.NodeId)
 	opened := node.mount.getOpenedFile(input.Fh)
 	return opened.dir.ReadDirPlus(cancel, input, out)
