diff --git fs/api.go fs/api.go
index 3cbd514..e8bcc3c 100644
--- fs/api.go
+++ fs/api.go
@@ -805,7 +805,7 @@ type Options struct {
 
 	// ServerCallbacks are optional callbacks to stub out notification functions
 	// for testing a filesystem without mounting it.
-	ServerCallbacks ServerCallbacks
+	ServerCallbacks fuse.ServerCallbacks
 
 	// Logger is a sink for diagnostic messages. Diagnostic
 	// messages are printed under conditions where we cannot
diff --git fs/bridge.go fs/bridge.go
index 5341bc3..d4747a6 100644
--- fs/bridge.go
+++ fs/bridge.go
@@ -54,17 +54,6 @@ type fileEntry struct {
 	wg sync.WaitGroup
 }
 
-// ServerCallbacks are calls into the kernel to manipulate the inode,
-// entry and page cache.  They are stubbed so filesystems can be
-// unittested without mounting them.
-type ServerCallbacks interface {
-	DeleteNotify(parent uint64, child uint64, name string) fuse.Status
-	EntryNotify(parent uint64, name string) fuse.Status
-	InodeNotify(node uint64, off int64, length int64) fuse.Status
-	InodeRetrieveCache(node uint64, offset int64, dest []byte) (n int, st fuse.Status)
-	InodeNotifyStoreCache(node uint64, offset int64, data []byte) fuse.Status
-}
-
 // TODO: fold serverBackingFdCallbacks into ServerCallbacks and bump API version
 type serverBackingFdCallbacks interface {
 	RegisterBackingFd(*fuse.BackingMap) (int32, syscall.Errno)
@@ -74,7 +63,7 @@ type serverBackingFdCallbacks interface {
 type rawBridge struct {
 	options Options
 	root    *Inode
-	server  ServerCallbacks
+	server  fuse.ServerCallbacks
 
 	// mu protects the following data.  Locks for inodes must be
 	// taken before rawBridge.mu
@@ -1254,7 +1243,7 @@ func (b *rawBridge) StatFs(cancel <-chan struct{}, input *fuse.InHeader, out *fu
 	return fuse.OK
 }
 
-func (b *rawBridge) Init(s *fuse.Server) {
+func (b *rawBridge) Init(s fuse.ServerCallbacks) {
 	b.server = s
 }
 
diff --git fuse/api.go fuse/api.go
index 7efcf8d..4659a5b 100644
--- fuse/api.go
+++ fuse/api.go
@@ -362,6 +362,17 @@ type MountOptions struct {
 	ExtraCapabilities uint64
 }
 
+// ServerCallbacks are calls into the kernel to manipulate the inode,
+// entry and page cache.  They are stubbed so filesystems can be
+// unittested without mounting them.
+type ServerCallbacks interface {
+	DeleteNotify(parent uint64, child uint64, name string) Status
+	EntryNotify(parent uint64, name string) Status
+	InodeNotify(node uint64, off int64, length int64) Status
+	InodeRetrieveCache(node uint64, offset int64, dest []byte) (n int, st Status)
+	InodeNotifyStoreCache(node uint64, offset int64, data []byte) Status
+}
+
 // RawFileSystem is an interface close to the FUSE wire protocol.
 //
 // Unless you really know what you are doing, you should not implement
@@ -477,7 +488,7 @@ type RawFileSystem interface {
 	// This is called on processing the first request. The
 	// filesystem implementation can use the server argument to
 	// talk back to the kernel (through notify methods).
-	Init(*Server)
+	Init(ServerCallbacks)
 
 	// Called after processing the last request.
 	OnUnmount()
diff --git fuse/defaultraw.go fuse/defaultraw.go
index b15df5e..ad9d911 100644
--- fuse/defaultraw.go
+++ fuse/defaultraw.go
@@ -16,7 +16,7 @@ func NewDefaultRawFileSystem() RawFileSystem {
 
 type defaultRawFileSystem struct{}
 
-func (fs *defaultRawFileSystem) Init(*Server) {
+func (fs *defaultRawFileSystem) Init(ServerCallbacks) {
 }
 
 func (fs *defaultRawFileSystem) OnUnmount() {
diff --git fuse/nodefs/fsconnector.go fuse/nodefs/fsconnector.go
index 25945ff..351b802 100644
--- fuse/nodefs/fsconnector.go
+++ fuse/nodefs/fsconnector.go
@@ -29,7 +29,7 @@ type FileSystemConnector struct {
 	debug bool
 
 	// Callbacks for talking back to the kernel.
-	server *fuse.Server
+	server fuse.ServerCallbacks
 
 	// Translate between uint64 handles and *Inode.
 	inodeMap handleMap
@@ -81,7 +81,7 @@ func NewFileSystemConnector(root Node, opts *Options) (c *FileSystemConnector) {
 }
 
 // Server returns the fuse.Server that talking to the kernel.
-func (c *FileSystemConnector) Server() *fuse.Server {
+func (c *FileSystemConnector) Server() fuse.ServerCallbacks {
 	return c.server
 }
 
diff --git fuse/nodefs/fsops.go fuse/nodefs/fsops.go
index 18d8f8a..cfde5cd 100644
--- fuse/nodefs/fsops.go
+++ fuse/nodefs/fsops.go
@@ -56,7 +56,7 @@ func (c *rawBridge) String() string {
 	return name
 }
 
-func (c *rawBridge) Init(s *fuse.Server) {
+func (c *rawBridge) Init(s fuse.ServerCallbacks) {
 	c.server = s
 	c.rootNode.Node().OnMount((*FileSystemConnector)(c))
 }
diff --git fuse/test/cache_test.go fuse/test/cache_test.go
index 6e5de0c..195acac 100644
--- fuse/test/cache_test.go
+++ fuse/test/cache_test.go
@@ -142,7 +142,7 @@ func TestFopenKeepCache(t *testing.T) {
 		t.Fatalf("ReadFile: got %q, want cached %q", c, before)
 	}
 
-	if minor := pathfs.Connector().Server().KernelSettings().Minor; minor < 12 {
+	if minor := pathfs.Connector().Server().(*fuse.Server).KernelSettings().Minor; minor < 12 {
 		t.Skipf("protocol v%d has no notify support.", minor)
 	}
 
