diff --git fuse/api.go fuse/api.go
index 4659a5b..42e1142 100644
--- fuse/api.go
+++ fuse/api.go
@@ -301,6 +301,13 @@ type MountOptions struct {
 	// for more details.
 	SyncRead bool
 
+	// Let the kernel use the writeback cache mode, as opposed to
+	// write-through mode.
+	//
+	// See the following page for more details:
+	// https://www.kernel.org/doc/Documentation/filesystems/fuse-io.txt
+	EnableWritebackCache bool
+
 	// DirectMount, if set, makes go-fuse first attempt to use syscall.Mount instead of
 	// fusermount to mount the filesystem. This will not update /etc/mtab
 	// but might be needed if fusermount is not available.
diff --git fuse/opcode.go fuse/opcode.go
index 98bf8a9..e1c4128 100644
--- fuse/opcode.go
+++ fuse/opcode.go
@@ -111,6 +111,9 @@ func doInit(server *protocolServer, req *request) {
 	if server.opts.EnableSymlinkCaching {
 		kernelFlags |= CAP_CACHE_SYMLINKS
 	}
+	if server.opts.EnableWritebackCache {
+		server.kernelSettings.Flags |= CAP_WRITEBACK_CACHE
+	}
 	if server.opts.EnableAcl {
 		kernelFlags |= CAP_POSIX_ACL
 	}
